src/Util/MigrationRollbackBatch.php

```diff
@@ -0,0 +1,244 @@
+
+      $result = $this->rollback();
+
+      $context['message'] = $this->t('"@migration_id" has been processed', ['@migration_id' => $this->migration->id()]);
+      $this->messenger->addMessage(
+        $this->t('"@migration_id" has been processed', ['@migration_id' => $this->migration->id()])
+      );
+
+      $context['results']['status'] = MigrationInterface::RESULT_COMPLETED;
+      $context['sandbox']['total'] = 0;
+      $context['finished'] = 1;
+
+      return $result;
+    }
+    catch (\Exception $e) {
+      $this->handleException($e, FALSE);
+      $this->messenger->addError(
+        $this->t(
+          'Rollback encountered error while processing batch: @e.', ['@e' => $e]
+        )
+      );
+
+      return MigrationInterface::RESULT_FAILED;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function rollback() {
+    // Only begin the rollback operation if the migration is currently idle.
+    if ($this->migration->getStatus() !== MigrationInterface::STATUS_IDLE) {
+      $this->message->display($this->t(
+        'Migration @id is busy with another operation: @status',
+        ['@id' => $this->migration->id(), '@status' => $this->t($this->migration->getStatusLabel())]), 'error');
+      return MigrationInterface::RESULT_FAILED;
+    }
+
+    // Set the migration status to rolling back.
+    $this->migration->setStatus(MigrationInterface::STATUS_ROLLING_BACK);
+
+    $id_map = $this->getIdMap();
+    $destination = $this->migration->getDestinationPlugin();
+    $return = MigrationInterface::RESULT_COMPLETED;
+
+    $this->iterator->rewind();
+
+    while ($this->iterator->valid()) {
+      if ($this->iterator->current() === NULL) {
+        $this->message->display($this->t('Skipped processing due to null destination identifier.'));
+        $source_key = $id_map->currentSource();
+        $id_map->delete($source_key);
+        continue;
+      }else{
+        $this->getEventDispatcher()
+          ->dispatch(new MigrateRowDeleteEvent($this->migration, $this->iterator->current()), MigrateEvents::PRE_ROW_DELETE);
+        $destination->rollback($this->iterator->current());
+        $this->getEventDispatcher()
+          ->dispatch(new MigrateRowDeleteEvent($this->migration, $this->iterator->current()), MigrateEvents::POST_ROW_DELETE);
+        $id_map->deleteDestination($this->iterator->current());
+      }
+
+      // Check for memory exhaustion.
+      if (($return = $this->checkStatus()) != MigrationInterface::RESULT_COMPLETED) {
+        break;
+      }
+
+      // If anyone has requested we stop, return the requested result.
+      if ($this->migration->getStatus() == MigrationInterface::STATUS_STOPPING) {
+        $return = $this->migration->getInterruptionResult();
+        $this->migration->clearInterruptionResult();
+        break;
+      }
+
+      $this->iterator->next();
+    }
+
+    return $return;
+  }
```

Comment:
As presently implemented, this would all be dead code, no?

Reply:
I reverted my code change around individual row processing.